File: pysandwich/generated_error_codes.py
import pysandwich.proto.errors_pb2 as SandwichErrorProto
from pysandwich.error_base import SandwichException


class APIError(SandwichException):
    """APIError exception."""

    _ERRORS_MAP = {
        SandwichErrorProto.APIERROR_CONFIGURATION: {
            "msg": """Configuration error.""",
        },
        SandwichErrorProto.APIERROR_SOCKET: {
            "msg": """Socket error.""",
        },
        SandwichErrorProto.APIERROR_TUNNEL: {
            "msg": """Tunnel error.""",
        },
    }


class ConfigurationError(SandwichException):
    """ConfigurationError exception."""

    _ERRORS_MAP = {
        SandwichErrorProto.CONFIGURATIONERROR_INVALID_IMPLEMENTATION: {
            "msg": """The selected implementation is invalid.""",
        },
        SandwichErrorProto.CONFIGURATIONERROR_UNSUPPORTED_IMPLEMENTATION: {
            "msg": """The selected implementation isn&#39;t supported.""",
        },
        SandwichErrorProto.CONFIGURATIONERROR_INVALID: {
            "msg": """Invalid configuration.""",
        },
        SandwichErrorProto.CONFIGURATIONERROR_INVALID_LISTENER: {
            "msg": """Invalid Listener configuration.""",
        },
    }


class ProtobufError(SandwichException):
    """ProtobufError exception."""

    _ERRORS_MAP = {
        SandwichErrorProto.PROTOBUFERROR_EMPTY: {
            "msg": """The protobuf message is empty.""",
        },
        SandwichErrorProto.PROTOBUFERROR_TOO_BIG: {
            "msg": """The protobuf message is too large.""",
        },
        SandwichErrorProto.PROTOBUFERROR_PARSE_FAILED: {
            "msg": """Failed to parse the protobuf message.""",
        },
        SandwichErrorProto.PROTOBUFERROR_NULLPTR: {
            "msg": """A null pointer was supplied.
 This error is thrown by &#39;sandwich_context_new&#39;, when the given source
 buffer is a null pointer.""",
        },
        SandwichErrorProto.PROTOBUFERROR_INVALID_ARGUMENT: {
            "msg": """/ An invalid value was given.""",
        },
    }


class TLSConfigurationError(SandwichException):
    """TLSConfigurationError exception."""

    _ERRORS_MAP = {
        SandwichErrorProto.TLSCONFIGURATIONERROR_UNSUPPORTED_IMPLEMENTATION: {
            "msg": """The selected implementation isn&#39;t supported.""",
        },
        SandwichErrorProto.TLSCONFIGURATIONERROR_INVALID_CASE: {
            "msg": """The configuration case (client/server) isn&#39;t valid.""",
        },
        SandwichErrorProto.TLSCONFIGURATIONERROR_EMPTY: {
            "msg": """The configuration is empty.""",
        },
        SandwichErrorProto.TLSCONFIGURATIONERROR_UNSUPPORTED_PROTOCOL_VERSION: {
            "msg": """Unsupported protocol version error.""",
        },
        SandwichErrorProto.TLSCONFIGURATIONERROR_PRIVATE_KEY_INCONSISTENT_WITH_CERTIFICATE: {
            "msg": """/ Inconsistency between a private key and the corresponding certificate.""",
        },
        SandwichErrorProto.TLSCONFIGURATIONERROR_INVALID: {
            "msg": """Invalid configuration.""",
        },
    }


class CertificateError(SandwichException):
    """CertificateError exception."""

    _ERRORS_MAP = {
        SandwichErrorProto.CERTIFICATEERROR_MALFORMED: {
            "msg": """Malformed certificate.""",
        },
        SandwichErrorProto.CERTIFICATEERROR_EXPIRED: {
            "msg": """Expired certificate.""",
        },
        SandwichErrorProto.CERTIFICATEERROR_NOT_FOUND: {
            "msg": """Certificate not found.""",
        },
        SandwichErrorProto.CERTIFICATEERROR_UNKNOWN: {
            "msg": """Unknown error. Can wrap a SystemError.""",
        },
        SandwichErrorProto.CERTIFICATEERROR_UNSUPPORTED: {
            "msg": """Certificate not supported by the underlying implementation.""",
        },
    }


class PrivateKeyError(SandwichException):
    """PrivateKeyError exception."""

    _ERRORS_MAP = {
        SandwichErrorProto.PRIVATEKEYERROR_MALFORMED: {
            "msg": """Malformed private key.""",
        },
        SandwichErrorProto.PRIVATEKEYERROR_NOT_FOUND: {
            "msg": """Private key not found.""",
        },
        SandwichErrorProto.PRIVATEKEYERROR_UNKNOWN: {
            "msg": """Unknown error. Can wrap a SystemError.""",
        },
        SandwichErrorProto.PRIVATEKEYERROR_UNSUPPORTED: {
            "msg": """Certificate not supported by the underlying implementation.""",
        },
        SandwichErrorProto.PRIVATEKEYERROR_NOT_SERVER: {
            "msg": """Not a server configuration.""",
        },
    }


class ASN1Error(SandwichException):
    """ASN1Error exception."""

    _ERRORS_MAP = {
        SandwichErrorProto.ASN1ERROR_INVALID_FORMAT: {
            "msg": """Invalid format.""",
        },
        SandwichErrorProto.ASN1ERROR_MALFORMED: {
            "msg": """Malformed ASN.1 document.""",
        },
    }


class ALPNError(SandwichException):
    """ALPNError exception."""

    _ERRORS_MAP = {
        SandwichErrorProto.ALPNERROR_LENGTH_ERROR: {
            "msg": """Protocol length is longer than 255 bytes.""",
        },
        SandwichErrorProto.ALPNERROR_INVALID_STRING: {
            "msg": """Protocol contains &#39;\x00&#39; byte or invalid string.""",
        },
    }


class DataSourceError(SandwichException):
    """DataSourceError exception."""

    _ERRORS_MAP = {
        SandwichErrorProto.DATASOURCEERROR_EMPTY: {
            "msg": """Empty data source.""",
        },
        SandwichErrorProto.DATASOURCEERROR_INVALID_CASE: {
            "msg": """Invalid case for data source.""",
        },
        SandwichErrorProto.DATASOURCEERROR_NOT_FOUND: {
            "msg": """Data not found on local filesystem.""",
        },
    }


class KEMError(SandwichException):
    """KEMError exception."""

    _ERRORS_MAP = {
        SandwichErrorProto.KEMERROR_INVALID: {
            "msg": """Invalid or unsupported KEM.""",
        },
        SandwichErrorProto.KEMERROR_TOO_MANY: {
            "msg": """Too many KEMs.""",
        },
    }


class SystemError(SandwichException):
    """SystemError exception."""

    _ERRORS_MAP = {
        SandwichErrorProto.SYSTEMERROR_MEMORY: {
            "msg": """Memory error (e.g. allocation failed).""",
        },
        SandwichErrorProto.SYSTEMERROR_INTEGER_OVERFLOW: {
            "msg": """Integer overflow.""",
        },
    }


class SocketError(SandwichException):
    """SocketError exception."""

    _ERRORS_MAP = {
        SandwichErrorProto.SOCKETERROR_BAD_FD: {
            "msg": """Bad file descriptor.""",
        },
        SandwichErrorProto.SOCKETERROR_CREATION_FAILED: {
            "msg": """Socket creation failed.""",
        },
        SandwichErrorProto.SOCKETERROR_BAD_NETADDR: {
            "msg": """Invalid network address.""",
        },
        SandwichErrorProto.SOCKETERROR_NETADDR_UNKNOWN: {
            "msg": """Failed to resolve network address.""",
        },
        SandwichErrorProto.SOCKETERROR_FSTAT_FAILED: {
            "msg": """Syscall &#39;fstat&#39; failed.""",
        },
        SandwichErrorProto.SOCKETERROR_NOT_SOCK: {
            "msg": """File descriptor is not a socket.""",
        },
        SandwichErrorProto.SOCKETERROR_GETSOCKNAME_FAILED: {
            "msg": """Syscall getsockname failed.""",
        },
        SandwichErrorProto.SOCKETERROR_SETSOCKOPT_FAILED: {
            "msg": """Syscall setsockopt failed.""",
        },
        SandwichErrorProto.SOCKETERROR_INVALID_AI_FAMILY: {
            "msg": """Invalid AI family.""",
        },
    }


class HandshakeError(SandwichException):
    """HandshakeError exception."""

    _ERRORS_MAP = {
        SandwichErrorProto.HANDSHAKEERROR_INVALID_SERVER_NAME: {
            "msg": """Invalid Server Name.""",
        },
        SandwichErrorProto.HANDSHAKEERROR_CERTIFICATE_VERIFICATION_FAILED: {
            "msg": """Certficate verification failed.""",
        },
        SandwichErrorProto.HANDSHAKEERROR_CERTIFICATE_EXPIRED: {
            "msg": """Certificate has expired.""",
        },
        SandwichErrorProto.HANDSHAKEERROR_CERTIFICATE_REVOKED: {
            "msg": """Certificate was revoked.""",
        },
        SandwichErrorProto.HANDSHAKEERROR_INVALID_CERTIFICATE: {
            "msg": """Invalid Certificate.""",
        },
        SandwichErrorProto.HANDSHAKEERROR_CERTIFICATE_SIGNATURE_VERIFICATION_FAILED: {
            "msg": """Signature verification error.""",
        },
        SandwichErrorProto.HANDSHAKEERROR_UNKNOWN_ERROR: {
            "msg": """Unknown handshake error.""",
        },
    }


class TunnelError(SandwichException):
    """TunnelError exception."""

    _ERRORS_MAP = {
        SandwichErrorProto.TUNNELERROR_INVALID: {
            "msg": """Invalid tunnel configuration.""",
        },
        SandwichErrorProto.TUNNELERROR_VERIFIER: {
            "msg": """Invalid tunnel verifier.""",
        },
        SandwichErrorProto.TUNNELERROR_UNKNOWN: {
            "msg": """Unknown error.""",
        },
    }


_ERROR_KIND_MAP = {
    SandwichErrorProto.ERRORKIND_API: APIError,
    SandwichErrorProto.ERRORKIND_CONFIGURATION: ConfigurationError,
    SandwichErrorProto.ERRORKIND_PROTOBUF: ProtobufError,
    SandwichErrorProto.ERRORKIND_TLS_CONFIGURATION: TLSConfigurationError,
    SandwichErrorProto.ERRORKIND_CERTIFICATE: CertificateError,
    SandwichErrorProto.ERRORKIND_PRIVATE_KEY: PrivateKeyError,
    SandwichErrorProto.ERRORKIND_ASN1: ASN1Error,
    SandwichErrorProto.ERRORKIND_ALPN: ALPNError,
    SandwichErrorProto.ERRORKIND_DATA_SOURCE: DataSourceError,
    SandwichErrorProto.ERRORKIND_KEM: KEMError,
    SandwichErrorProto.ERRORKIND_SYSTEM: SystemError,
    SandwichErrorProto.ERRORKIND_SOCKET: SocketError,
    SandwichErrorProto.ERRORKIND_HANDSHAKE: HandshakeError,
    SandwichErrorProto.ERRORKIND_TUNNEL: TunnelError,
}

File: rust/error/generated_error_codes.rs
GenProtoBasedErrorCode!({
      kind: ERRORKIND_API,
      sym: APIError,
      desc: "API errors.
 The following errors can occur during a call to the Context API.",
      values: [
        APIERROR_CONFIGURATION => "Configuration error.",
APIERROR_SOCKET => "Socket error.",
APIERROR_TUNNEL => "Tunnel error.",

      ],
  },
{
      kind: ERRORKIND_CONFIGURATION,
      sym: ConfigurationError,
      desc: "Errors regarding configurations.",
      values: [
        CONFIGURATIONERROR_INVALID_IMPLEMENTATION => "The selected implementation is invalid.",
CONFIGURATIONERROR_UNSUPPORTED_IMPLEMENTATION => "The selected implementation isn&#39;t supported.",
CONFIGURATIONERROR_INVALID => "Invalid configuration.",
CONFIGURATIONERROR_INVALID_LISTENER => "Invalid Listener configuration.",

      ],
  },
{
      kind: ERRORKIND_PROTOBUF,
      sym: ProtobufError,
      desc: "Errors regarding protobuf.",
      values: [
        PROTOBUFERROR_EMPTY => "The protobuf message is empty.",
PROTOBUFERROR_TOO_BIG => "The protobuf message is too large.",
PROTOBUFERROR_PARSE_FAILED => "Failed to parse the protobuf message.",
PROTOBUFERROR_NULLPTR => "A null pointer was supplied.
 This error is thrown by &#39;sandwich_context_new&#39;, when the given source
 buffer is a null pointer.",
PROTOBUFERROR_INVALID_ARGUMENT => "/ An invalid value was given.",

      ],
  },
{
      kind: ERRORKIND_TLS_CONFIGURATION,
      sym: TLSConfigurationError,
      desc: "Errors regarding TLS configurations.",
      values: [
        TLSCONFIGURATIONERROR_UNSUPPORTED_IMPLEMENTATION => "The selected implementation isn&#39;t supported.",
TLSCONFIGURATIONERROR_INVALID_CASE => "The configuration case (client/server) isn&#39;t valid.",
TLSCONFIGURATIONERROR_EMPTY => "The configuration is empty.",
TLSCONFIGURATIONERROR_UNSUPPORTED_PROTOCOL_VERSION => "Unsupported protocol version error.",
TLSCONFIGURATIONERROR_PRIVATE_KEY_INCONSISTENT_WITH_CERTIFICATE => "Inconsistency between a private key and the corresponding certificate.",
TLSCONFIGURATIONERROR_UNSUPPORTED_CIPHERSUITE => "Unsupported TLS ciphersuite error.",
TLSCONFIGURATIONERROR_UNSUPPORTED_CONTROL_CHARACTERS => "Control characters are not allowed.",
TLSCONFIGURATIONERROR_INVALID => "Invalid configuration.",

      ],
  },
{
      kind: ERRORKIND_CERTIFICATE,
      sym: CertificateError,
      desc: "Certificate errors.",
      values: [
        CERTIFICATEERROR_MALFORMED => "Malformed certificate.",
CERTIFICATEERROR_EXPIRED => "Expired certificate.",
CERTIFICATEERROR_NOT_FOUND => "Certificate not found.",
CERTIFICATEERROR_UNKNOWN => "Unknown error. Can wrap a SystemError.",
CERTIFICATEERROR_UNSUPPORTED => "Certificate not supported by the underlying implementation.",

      ],
  },
{
      kind: ERRORKIND_PRIVATE_KEY,
      sym: PrivateKeyError,
      desc: "Private key errors.",
      values: [
        PRIVATEKEYERROR_MALFORMED => "Malformed private key.",
PRIVATEKEYERROR_NOT_FOUND => "Private key not found.",
PRIVATEKEYERROR_UNKNOWN => "Unknown error. Can wrap a SystemError.",
PRIVATEKEYERROR_UNSUPPORTED => "Certificate not supported by the underlying implementation.",
PRIVATEKEYERROR_NOT_SERVER => "Not a server configuration.",

      ],
  },
{
      kind: ERRORKIND_ASN1,
      sym: ASN1Error,
      desc: "ASN.1 errors.",
      values: [
        ASN1ERROR_INVALID_FORMAT => "Invalid format.",
ASN1ERROR_MALFORMED => "Malformed ASN.1 document.",

      ],
  },
{
      kind: ERRORKIND_ALPN,
      sym: ALPNError,
      desc: "ALPN errors.",
      values: [
        ALPNERROR_LENGTH_ERROR => "Protocol length is longer than 255 bytes.",
ALPNERROR_INVALID_STRING => "Protocol contains &#39;\x00&#39; byte or invalid string.",

      ],
  },
{
      kind: ERRORKIND_DATA_SOURCE,
      sym: DataSourceError,
      desc: "DataSource errors.",
      values: [
        DATASOURCEERROR_EMPTY => "Empty data source.",
DATASOURCEERROR_INVALID_CASE => "Invalid case for data source.",
DATASOURCEERROR_NOT_FOUND => "Data not found on local filesystem.",

      ],
  },
{
      kind: ERRORKIND_KEM,
      sym: KEMError,
      desc: "KEM errors.",
      values: [
        KEMERROR_INVALID => "Invalid or unsupported KEM.",
KEMERROR_TOO_MANY => "Too many KEMs.",

      ],
  },
{
      kind: ERRORKIND_SYSTEM,
      sym: SystemError,
      desc: "System errors.",
      values: [
        SYSTEMERROR_MEMORY => "Memory error (e.g. allocation failed).",
SYSTEMERROR_INTEGER_OVERFLOW => "Integer overflow.",
SYSTEMERROR_BACKEND => "Backend error.",

      ],
  },
{
      kind: ERRORKIND_SOCKET,
      sym: SocketError,
      desc: "Socket errors.
 These errors are used in io/socket.",
      values: [
        SOCKETERROR_BAD_FD => "Bad file descriptor.",
SOCKETERROR_CREATION_FAILED => "Socket creation failed.",
SOCKETERROR_BAD_NETADDR => "Invalid network address.",
SOCKETERROR_NETADDR_UNKNOWN => "Failed to resolve network address.",
SOCKETERROR_FSTAT_FAILED => "Syscall &#39;fstat&#39; failed.",
SOCKETERROR_NOT_SOCK => "File descriptor is not a socket.",
SOCKETERROR_GETSOCKNAME_FAILED => "Syscall getsockname failed.",
SOCKETERROR_SETSOCKOPT_FAILED => "Syscall setsockopt failed.",
SOCKETERROR_INVALID_AI_FAMILY => "Invalid AI family.",

      ],
  },
{
      kind: ERRORKIND_HANDSHAKE,
      sym: HandshakeError,
      desc: "",
      values: [
        HANDSHAKEERROR_INVALID_SERVER_NAME => "Invalid Server Name.",
HANDSHAKEERROR_CERTIFICATE_VERIFICATION_FAILED => "Certficate verification failed.",
HANDSHAKEERROR_CERTIFICATE_EXPIRED => "Certificate has expired.",
HANDSHAKEERROR_CERTIFICATE_REVOKED => "Certificate was revoked.",
HANDSHAKEERROR_INVALID_CERTIFICATE => "Invalid Certificate.",
HANDSHAKEERROR_CERTIFICATE_SIGNATURE_VERIFICATION_FAILED => "Signature verification error.",
HANDSHAKEERROR_DEPTH_EXCEEDED => "Certificate chain too long or pathlen exceeded.",
HANDSHAKEERROR_UNSUPPORTED_PROTOCOL => "Unsupported protocol.",
HANDSHAKEERROR_NO_SHARED_CIPHER => "No shared cipher.",
HANDSHAKEERROR_NO_SUITABLE_KEY_SHARE => "No suitable key share.",
HANDSHAKEERROR_UNKNOWN_ERROR => "Unknown handshake error.",

      ],
  },
{
      kind: ERRORKIND_TUNNEL,
      sym: TunnelError,
      desc: "Tunnel error.",
      values: [
        TUNNELERROR_INVALID => "Invalid tunnel configuration.",
TUNNELERROR_VERIFIER => "Invalid tunnel verifier.",
TUNNELERROR_UNKNOWN => "Unknown error.",

      ],
  },
);
